### PR TITLE
FIX Handle getting HasOneRelationFieldInterface passed as an array

### PR DESCRIPTION
**Pull request by emteknetnz**
Issue https://github.com/silverstripe/silverstripe-elemental/issues/1155

This fixes bugs with autoscaffolded SearchableDropdownField for has_one SiteTree relations

Replication steps

- Create a non-inline editable elemental content block
- Add a has_one relation to a 'MySiteTree' => SiteTree::class
- Add include the validation php block below

```php
    public function getCMSCompositeValidator(): CompositeValidator
    {
        return CompositeValidator::create([RequiredFields::create(['MySiteTreeID'])]);
    }
```

You can leave the field blank and submit and the required fields validation won't trigger

### CODE REVIEW

**Comment by GuySartorelli** [comment]
I think the validator is working correctly - the developer should use the correct field name when setting their validator, i.e. it should be 
```php
RequiredFields::create(['MySiteTreeID'])
```
and **not**

```php
RequiredFields::create(['MySiteTree'])
```

File: tests/php/Forms/RequiredFieldsTest.php [hunk 1]
```diff
@@ -289,17 +290,27 @@ public function testFieldIsRequired()
...
-        $form->Fields()->push($field);
-        $validator = new RequiredFields('TestField');
-        $validator->setForm($form);
-        // blank string and '0' are fail required field validation
-        $this->assertFalse($validator->php(['TestField' => '']));
-        $this->assertFalse($validator->php(['TestField' => '0']));
-        // '1' passes required field validation
-        $this->assertTrue($validator->php(['TestField' => '1']));
+        $fieldClasses = [TreeDropdownField::class, SearchableDropdownField::class];
+        $suffixes = ['', 'ID'];
```
Comment (GuySartorelli) [inline tests/php/Forms/RequiredFieldsTest.php]: Add these as scenarios in a dataprovider

File: src/Forms/RequiredFields.php [hunk 2]
```diff
@@ -104,16 +104,29 @@ public function php($data)
...
             }
 
             // submitted data for file upload fields come back as an array
-            $value = isset($data[$fieldName]) ? $data[$fieldName] : null;
+            $value = null;
+            if (isset($data[$fieldName])) {
+                $value = $data[$fieldName];
+            } elseif (isset($data[$fieldName . 'ID'])) {
+                $value = $data[$fieldName . 'ID'];
+            }
```
Comment (GuySartorelli) [inline src/Forms/RequiredFields.php]: The comment is about file upload fields, but you're doing something here with the `ID` suffix which is clearly more about `has_one` relations. I think the comment was meant to be above the `is_array` check below? But that's also not only about file fields.

Please replace this comment with something about checking for `ID` suffixes. Please also only add the suffix for fields which implement `HasOneRelationFieldInterface`.

File: src/Forms/RequiredFields.php [hunk 3]
```diff
@@ -104,16 +104,29 @@ public function php($data)
                 $fieldName = $fieldName->getName();
             } else {
                 $formField = $fields->dataFieldByName($fieldName);
+                if (is_null($formField)) {
+                    $formField = $fields->dataFieldByName($fieldName . 'ID');
+                }
```
Comment (GuySartorelli) [inline src/Forms/RequiredFields.php]: If the form field doesn't implement `HasOneRelationFieldInterface`, we should set the variable back to `null` - this logic should only be done for `has_one` fields.

File: src/Forms/RequiredFields.php [hunk 4]
```diff
@@ -104,16 +104,29 @@ public function php($data)
...
                 if ($formField instanceof FileField && isset($value['error']) && $value['error']) {
                     $error = true;
                 } else {
-                    $error = (count($value ?? [])) ? false : true;
+                    if (is_a($formField, HasOneRelationFieldInterface::class) && isset($value['value'])) {
+                        $stringValue = (string) $value['value'];
+                        $error = in_array($stringValue, ['0', '']);
+                    } else {
+                        $error = (count($value ?? [])) ? false : true;
+                    }
```
Comment (GuySartorelli) [inline src/Forms/RequiredFields.php:121]: This doesn't seem related to the commit message. It also doesn't seem necessary? Can you please provide steps to test this? Validation seems to work without it.

**Review by GuySartorelli — CHANGES_REQUESTED** [review]
I'll treat this PR as an enhancement instead of a bugfix, since I think the current behaviour isn't buggy, even though it's potentially confusing since some fields use the `ID` suffix and some don't.

To that end, please update the commit prefix to `ENH`.

File: src/Forms/RequiredFields.php (hunk 4 above)
Comment (emteknetnz) [reply src/Forms/RequiredFields.php:121]: Have updated commit message. I've removed the 'ID' suffix bits from this PR.

I've kept the commit prefix as 'FIX' because it's fixing behaviour that should work though doesn't.  Previously ENH did make sense because I was including adding the 'ID' suffix, though I've now removed that.

File: tests/php/Forms/RequiredFieldsTest.php (hunk 1 above)
Comment (emteknetnz) [reply tests/php/Forms/RequiredFieldsTest.php]: Updated

File: src/Forms/RequiredFields.php (hunk 2 above)
Comment (emteknetnz) [reply src/Forms/RequiredFields.php]: Updated

File: src/Forms/RequiredFields.php (hunk 3 above)
Comment (emteknetnz) [reply src/Forms/RequiredFields.php]: Have removed code

**Comment by GuySartorelli** [comment]
Does the bug you're fixing affect CMS 5.2?
If so, please retarget to the `5.2` branch since it's fixing a bug

**Review by GuySartorelli — COMMENTED** [review]
I've tried to reproduce the bug mentioned in the description both from the `5` and `5.2` branch - I can't reproduce it from either. Validation works fine without this PR in both scenarios.

Please try spinning up a fresh installation and testing your own reproduction steps.

**Comment by emteknetnz** [comment]
Yeah looks like we don't need this one, I think there may have been some initial confusion with the "ID" suffix making me think things were broken when they were not

**Comment by emteknetnz** [comment]
Turns out we do need this one, it's required for doing a page save of an inline-block where the has_one SiteTree::class in in RequiredFields.  It's NOT required for saving non-inline block's.

It showed up as a behat failure in page-save-validation.feature on https://github.com/silverstripe/silverstripe-elemental/actions/runs/8961905824/job/24613378755?pr=1178#step:12:1052

**Review by GuySartorelli — APPROVED** [review]
It turns out there's a difference with some fields between "I have not selected anything" and "I have specifically clicked on the dropdown and selected the empty option".

Since form submission already correctly handles both cases, and it's been like that for donkey's years, I think it makes sense to handle this here as you've done.
